### ISSUE TITLE
Automate releases

### ISSUE THREAD

**Issue by mattwynne** [cucumber/common#1688]
Using this issue to bring together the efforts scattered across other repos.

The goal is to have all releases to RubyGems, NPM, Maven etc. made by GitHub Actions workflows. We'll use GitHub Environments to secure the secrets used to publish packages.

- [x] cucumber/build https://github.com/cucumber/build/issues/42
- [x] cucumber/cucumber-ruby-wire https://github.com/cucumber/cucumber-ruby-wire/pull/49
- [x] factor out a `publish-rubygem` action
- [x] factor out a `versions` action to read the changelog latest and tagged release versions
- [x] factor out a `create-release` action. WIP in https://github.com/cucumber-actions/create-release
- [x] review all instances of release workflows and ensure they're using the latest versions of our actions
- [x] archive `get-released-version` action which is superseded by the new `versions` action.
- [x] cucumber/cucumber-expressions https://github.com/cucumber/common/issues/1688
- [x] cucumber-ruby-core: https://github.com/cucumber/cucumber-ruby-core/pull/224
- [x] cucumber-ruby: https://github.com/cucumber/cucumber-ruby/pull/1567
- [ ] cucumber-parent: https://github.com/cucumber/cucumber-parent/issues/32
- [ ] create a `publish-go` action
- [ ] create a `check-release-commit-is-on-main` action (see [here](https://github.com/cucumber/cucumber-ruby-core/pull/224#discussion_r718838439)
- [ ] factor out RELEASING.md to try having a single one compatible with all the repos
  - [ ] script or automate the bump of the version numbers into package descriptors
  - [x] extract update of gemspec of the release process, or document it as a generic process as part of the release one

**Comment by mattwynne** [comment]
The current strategy to mark the commit to be released with a temporary `pre-release/*` branch is a bit fragile as, if that branch is deleted before the release is made, we don't know what commit to tag.

Maybe the pre-release script needs to make a "Draft" Release (and a tag), then we can just publish that Release when we make the actual release.

**Comment by mattwynne** [comment]
We can pull out the sha of the commit being released like this:

```
url=$(gh pr list --state closed --base release/v0.8.0 --json url | jq -r '.[0].url')
sha=$(gh pr view $url --json commits --jq '.commits[-1].oid')
```

That should work even if the `pre-release/*` branch has been deleted.

**Comment by mattwynne** [comment]
I've sketched out a repo for a `create-release` action where we can properly test the different scenarios around tagging release commits. It's not finished yet. https://github.com/cucumber-actions/create-release

**Comment by aslakhellesoy** [comment]
We now have a working GitHub Actions release workflow in `cucumber/cucumber-expressions`

[Preparing a release](https://github.com/cucumber/cucumber-expressions/blob/main/RELEASING.md#preparing-a-release) is still a bit too manual for my liking. Things I'd like to improve: 

- [ ] Make it easier for people to install the [changelog](https://github.com/cucumber/changelog) tool. We can improve the existing [goreleaser](https://goreleaser.com/) config so that people can install it with various package managers.
- [ ] Provide a `scripts/prepare-release script` that runs `changelog release ...`, updates the package descriptors and pushes to github.

Now that we're splitting up `common` into more and more repos, I think each repo will need a `RELEASING.md` file and a `scripts/prerelease` script. I also think they will be the same. So we should look into a solution where we can maintain common files in a single repo and easily sync them to select repos.

**Comment by aurelien-reeves** [comment]
I must say that once it is working, I find the release process pretty smooth. Like it a lot.

I see two things we could improve:
- we could very easily push the release commit only on the release branch and never on the main branch: maybe we could make sure the commit we are about to release exists on the main branch
- we could very easily too release a non working module. I think we should run the tests before the publication of the modules, even if tests have already been executed on the main branch. Actually nothing prevent us to push non working stuff on the release branch right now

**Comment by mattwynne** [comment]
>     * Provide a `scripts/prepare-release` script that runs `changelog release ...`, updates the package descriptors and pushes to github.

`gh` lets us trigger `workflow_dispatch` events to run arbitrary workflows with parameters, so we could probably bundle this script as a GH action and then use something like this to trigger it:

    gh workflow run pre-release.yaml -f next-version=2.0.0

WDYT?

**Comment by aurelien-reeves** [comment]
What would be the value to use a workflow over a local script?

**Comment by aslakhellesoy** [comment]
The main benefit would be that we don’t have to install the `changelog` script locally.

we also wouldn’t have to copy-paste the pre-release script across multiple repos (but we’d still have to enable the action for each repo)

**Comment by aurelien-reeves** [comment]
I have added some TODOs too following the comments here: https://github.com/cucumber/cucumber-ruby-core/pull/224#issuecomment-931954367

**Comment by aurelien-reeves** [comment]
Some feedback after having released `cucumber-ruby-core`, `cucumber-ruby-wire` and `cucumber-ruby` using the new release process, the one with a single `push` on a dedicated `release/vX.Y.Z` branch.

First of all, it worked as expected. It was pretty smooth 🤩

In short: 
- we prepare the release as usual (preparation of the changelog, update of the gemspec, ...) 
- push a "Release" commit to main
- then we can push the main branch to a dedicated `release/vX.Y.Z` branch. That will trigger a dedicated workflow which will release the new version.

The workflow has been copied from `cucumber-expressions`, and it has been tweaked:

- I have added a check to make sure what we were about to release was pushed on `main` before (https://github.com/cucumber/cucumber-ruby/blob/main/.github/workflows/release.yml#L10)
- I have added the execution of the tests before publishing the gem (https://github.com/cucumber/cucumber-ruby/blob/main/.github/workflows/release.yml#L22). That may slow-down the process, but it make it far more safer.

Also, as the repos have a single implementation, there was a single release workflow with the publication of the ruby gem, then the creation of the release on github. That may not be as simple with repos with multiple languages.

![image](https://user-images.githubusercontent.com/24386780/137096001-1268ab99-b6d8-474a-a969-c0d18c4852a2.png)

As an attempt of factorization, I have created an action to execute the tests, and use that action as part of the test workflow and the release workflow.

I don't know how far we could factorize the process. There are some redundancy across the repositories, but I am not sure this is because of duplication. We could maybe create a precheck action, but other jobs are already pretty well factorized into their own dedicated actions. What we have to take care of is: if we tweak the release for a repository, we may have to update it for other repositories as well.

Also as @aslakhellesoy [pointed out](https://github.com/cucumber/cucumber-ruby/pull/1578#pullrequestreview-777396794), the RELEASING.md document could be standardized.

I am pretty confident with that process. BUT, there is a point we have to be extra careful: the security of the process rely on github settings: the protection of `release/*` branches, and the associated "Release" environment. If an option in those settings is missed, some secrets could be compromised.
